src/modules/roc_core/target_posix/roc_core/cmdline_parser.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+//! @file roc_core/target_posix/roc_core/cmdline_parser.h
+//! @brief Command line parser.
+
+#ifndef ROC_CORE_CMDLINE_PARSER_H_
+#define ROC_CORE_CMDLINE_PARSER_H_
+
+#include "roc_core/stddefs.h"
+#include "roc_core/noncopyable.h"
+
+namespace roc {
+namespace core {
+
+//! Command line parser template.
+//!
+//! @tparam T represents a type for a container that holds parsed arguments.
+//! @tparam ParseFunc represents a function to parse command line arguments.
+//! @tparam DestroyFunc represents a function to release any resources
+//! allocated during the parsing.
+template
+    <
+      class T
+    , class ParseFunc = int(*)(int, char**, T*)
+    , class DestroyFunc = void(*)(T*)
+    >
+class CmdlineParser : public core::NonCopyable<> {
```

Comment:
The class looks odd. This "parser" does not parse anything. It actually implements RAII on top of the two generic functions that are not specific to parsing.

I suggest to add a simpler helper (to roc_core) which solves the problem directly:

```
template <class T, void (*Func)(T)> class ScopedDestructor : public NonCopyable<> {
public:
    ScopedDestructor(T obj)
        : obj_(obj) {
    }

    ~ScopedDestructor() {
        Func(obj_);
    }

private:
    T obj_;
};
```

And then use it like this:

```
    gengetopt_args_info args;

    const int code = cmdline_parser(argc, argv, &args);
    if (code != 0) {
        return code;
    }

    core::ScopedDestructor<gengetopt_args_info*, cmdline_parser_free> args_destructor(
        &args);
```

You could also reuse core::UniquePtr by providing it a custom Destroyer class, but this would require too much boilerplate code, I think.

Reply:
Indeed, it is. My first solution was similar to your and I also tried to
provide the Destoyer to the core::UniquePtr. I still don't like both approaches
because you separate the parse operation and wrapping object into RAII, which
ultimetely does not stop from just forgetting to wrap the parsed object.

Well, as your solution is more simple, we can use it.

Reply:
> you separate the parse operation and wrapping object into RAII

This is not good, but IMHO it's not critical at all; we have just a couple of command line tools, not dozens of them :)

Reply:
Agree, just wanted to clarify why this "parser" appeared.